lib_eio/net.ml

```diff
@@ -208,3 +212,26 @@ let getaddrinfo_datagram ?service t hostname =
 let getnameinfo (t:#t) sockaddr = t#getnameinfo sockaddr
 
 let close = Flow.close
+
+let rec try_connection env sw timeout host service = function
+  | [] -> raise @@ Connection_failure (host, service)
+  | (_x, ip, port)::addrs -> (
+    match Time.with_timeout_exn env#clock timeout (fun () -> connect ~sw env#net (`Tcp (ip, port))) with
+    | conn -> conn
+    | exception Unix_error(_,_,_)
+    | exception Time.Timeout -> try_connection env sw timeout host service addrs)
+
+let with_tcp_connect ?(timeout=0.1) ~host ~service env f =
+  let addrs =
+    getaddrinfo_stream ~service env#net host
+    |> List.filter_map (function
+        | `Tcp (x, port) -> Some (Ipaddr.fold ~v4:(fun ip -> `V4, ip, port) ~v6:(fun ip -> `V6, ip, port) x)
+        | `Unix _ -> None)
+    |> List.sort ( fun (a,_,_) (b,_,_) ->
```

Comment:
`getaddrinfo_stream` returns the results in the preferred order. Sorting them defeats that.

Reply:
IIUC `getaddrinfo_stream` doesn't sort the `Eio.Net.Ipaddr.t` in the correct order, i.e. `V6` address first. The following snippet illustrates this:
```
Eio_main.run @@ fun env ->
Eio.Net.getaddrinfo_stream ~service:"http" env#net "www.example.com"
|> List.iter (function
  | `Tcp (x, port) -> Eio.Net.Ipaddr.fold ~v4:(fun ip -> print_string "V4\n") ~v6:(fun ip -> print_string "V6\n") x
  | `Unix _ -> ());;
```
The above prints:
```
V4
V6
```
Additionally, `getaddrinfo_stream` doesn't filter out the `Unix _ ` bit. So the code takes care of that too.

Reply:
getaddrinfo(3) says:

> Normally, the application should try using the addresses in  the  order  in  which they are returned.

So if V4 addresses are coming first, then presumably that's what that system wants us to try first.

https://www.ietf.org/rfc/rfc3484.txt says:
> Well-behaved applications SHOULD iterate through the list of
>  addresses returned from getaddrinfo() until they find a working
>  address.

Reply:
Ah, thanks for the reference RFC. I see what you mean, now. Since, we are using `getaddrinfo` behind the scenes it makes sense to follow what the RFC says. Accordingly, I will remove the sorting functionality.

Just FYI, the mechanism adopted by the PR is more in line with the happy eyeballs RFC (https://www.rfc-editor.org/rfc/rfc8305). The RFC seems to suggest for the preference of IPv6 as opposed to IPv4 due to the former becoming more prevalent.

Reply:
Removed the sorting. It now uses the order returned by `getaddrinfo`.